File: frontend/src/api/shipping-tracking.ts
/**
 * Shipping Tracking API
 * Integrates with multiple shipping companies (UPS, Canada Express, Royal Mail, etc.)
 */

import { apiClient } from './client';

export interface ShipmentTracking {
  id: string;
  order_id: string;
  order_item_id?: string;
  tracking_number: string;
  carrier: string;
  status: string;
  shipment_type: string;
  shipped_at?: string;
  estimated_delivery?: string;
  actual_delivery?: string;
  current_location?: {
    city?: string;
    state?: string;
    country?: string;
    coordinates?: {
      lat: number;
      lng: number;
    };
  };
  delivery_instructions?: string;
  package_weight?: number;
  package_dimensions?: {
    length: number;
    width: number;
    height: number;
  };
  package_value?: number;
  insurance_amount?: number;
  service_level?: string;
  delivery_signature_required: boolean;
  external_tracking_url?: string;
  tracking_events: TrackingEvent[];
  created_at: string;
  updated_at: string;
}

export interface TrackingEvent {
  id: string;
  shipment_id: string;
  event_timestamp: string;
  event_type: string;
  event_description: string;
  event_location?: {
    city?: string;
    state?: string;
    country?: string;
    coordinates?: {
      lat: number;
      lng: number;
    };
  };
  carrier_event_code?: string;
  estimated_delivery?: string;
  delay_reason?: string;
  contact_name?: string;
  contact_phone?: string;
  source: string;
  created_at: string;
}

export interface ShippingProvider {
  id: string;
  name: string;
  carrier: string;
  api_url: string;
  tracking_url_template: string;
  is_active: boolean;
  created_at: string;
}

export interface ShipmentCreateRequest {
  order_id: string;
  order_item_id?: string;
  carrier: string;
  tracking_number: string;
  shipment_type?: string;
  origin_address?: any;
  destination_address?: any;
  delivery_instructions?: string;
  package_weight?: number;
  package_dimensions?: {
    length: number;
    width: number;
    height: number;
  };
  package_value?: number;
  insurance_amount?: number;
  service_level?: string;
  delivery_signature_required?: boolean;
  delivery_confirmation?: string;
  notes?: string;
  internal_notes?: string;
  shipped_at?: string;
}

export interface ShipmentUpdateRequest {
  status: string;
  event_description?: string;
  event_location?: any;
  contact_name?: string;
  contact_phone?: string;
}

export interface TrackingRequest {
  tracking_number: string;
  carrier: string;
}

export class ShippingTrackingAPI {
  /**
   * Create a new shipment tracking record
   */
  static async createShipment(shipmentData: ShipmentCreateRequest) {
    const response = await apiClient.post('/shipping-tracking/shipments', shipmentData);
    return response.data;
  }

  /**
   * Get detailed tracking information for a shipment
   */
  static async getShipmentTracking(shipmentId: string) {
    const response = await apiClient.get(`/shipping-tracking/shipments/${shipmentId}`);
    return response.data;
  }

  /**
   * Get all shipments for an order
   */
  static async getOrderShipments(orderId: string) {
    const response = await apiClient.get(`/shipping-tracking/orders/${orderId}/shipments`);
    return response.data;
  }

  /**
   * Track a shipment using carrier-specific integration
   */
  static async trackShipment(trackingRequest: TrackingRequest) {
    const response = await apiClient.post('/shipping-tracking/track', trackingRequest);
    return response.data;
  }

  /**
   * Update shipment status and create tracking event
   */
  static async updateShipmentStatus(shipmentId: string, updateData: ShipmentUpdateRequest) {
    const response = await apiClient.put(`/shipping-tracking/shipments/${shipmentId}/status`, updateData);
    return response.data;
  }

  /**
   * Get list of supported shipping carriers
   */
  static async getSupportedCarriers() {
    const response = await apiClient.get('/shipping-tracking/carriers');
    return response.data;
  }

  /**
   * Get tracking URL for a shipment
   */
  static getTrackingUrl(shipment: ShipmentTracking): string | null {
    if (shipment.external_tracking_url) {
      return shipment.external_tracking_url;
    }
    
    // Fallback: generate carrier-specific URL if template is available
    const carrierUrls: Record<string, string> = {
      'ups': `https://www.ups.com/track?tracknum=${shipment.tracking_number}`,
      'canada_express': `https://www.canadapost.ca/track?trackingNumber=${shipment.tracking_number}`,
      'royal_mail': `https://www.royalmail.com/track-your-item?trackNumber=${shipment.tracking_number}`,
      'fedex': `https://www.fedex.com/fedextrack/?trknbr=${shipment.tracking_number}`,
      'dhl': `https://www.dhl.com/en/express/tracking.html?AWB=${shipment.tracking_number}`,
      'usps': `https://tools.usps.com/go/TrackConfirmAction_input?qtc_tLabels1=${shipment.tracking_number}`,
      'canada_post': `https://www.canadapost.ca/cpotools/track/trackPackage?trackingNumber=${shipment.tracking_number}`,
      'purolator': `https://www.purolator.com/en/track/tracking-summary.xhtml?pin=${shipment.tracking_number}`
    };
    
    return carrierUrls[shipment.carrier] || null;
  }

  /**
   * Get carrier-specific tracking URL template
   */
  static getCarrierTrackingUrl(carrier: string, trackingNumber: string): string {
    const templates: Record<string, string> = {
      'ups': `https://www.ups.com/track?tracknum={tracking_number}`,
      'canada_express': `https://www.canadapost.ca/track?trackingNumber={tracking_number}`,
      'royal_mail': `https://www.royalmail.com/track-your-item?trackNumber={tracking_number}`,
      'fedex': `https://www.fedex.com/fedextrack/?trknbr={tracking_number}`,
      'dhl': `https://www.dhl.com/en/express/tracking.html?AWB={tracking_number}`,
      'usps': `https://tools.usps.com/go/TrackConfirmAction_input?qtc_tLabels1={tracking_number}`,
      'canada_post': `https://www.canadapost.ca/cpotools/track/trackPackage?trackingNumber={tracking_number}`,
      'purolator': `https://www.purolator.com/en/track/tracking-summary.xhtml?pin={tracking_number}`
    };
    
    return templates[carrier] || `https://www.google.com/search?q=${carrier}+tracking+${trackingNumber}`;
  }

  /**
   * Get carrier logo URL
   */
  static getCarrierLogo(carrier: string): string {
    const logos: Record<string, string> = {
      'ups': 'https://upload.wikimedia.org/wikipedia/commons/thumb/8/83/UPS_logo_2014.svg/200px-UPS_logo_2014.svg.png',
      'canada_express': 'https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Canada_Post_logo_2020.svg/200px-Canada_Post_logo_2020.svg.png',
      'royal_mail': 'https://upload.wikimedia.org/wikipedia/commons/thumb/0/06/Royal_Mail_logo.svg/200px-Royal_Mail_logo.svg.png',
      'fedex': 'https://upload.wikimedia.org/wikipedia/commons/thumb/6/69/FedEx_logo_2014.svg/200px-FedEx_logo_2014.svg.png',
      'dhl': 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/4d/DHL_2017_logo.svg/200px-DHL_2017_logo.svg.png',
      'usps': 'https://upload.wikimedia.org/wikipedia/commons/thumb/e/ee/USPS_logo_2014.svg/200px-USPS_logo_2014.svg.png',
      'canada_post': 'https://upload.wikimedia.org/wikipedia/commons/thumb/6/6d/Canada_Post_logo_2020.svg/200px-Canada_Post_logo_2020.svg.png',
      'purolator': 'https://upload.wikimedia.org/wikipedia/commons/thumb/3/3f/Purolator_logo.svg/200px-Purolator_logo.svg.png'
    };
    
    return logos[carrier] || '/images/default-carrier-logo.png';
  }

  /**
   * Get carrier display name
   */
  static getCarrierDisplayName(carrier: string): string {
    const names: Record<string, string> = {
      'ups': 'UPS',
      'canada_express': 'Canada Express',
      'royal_mail': 'Royal Mail',
      'fedex': 'FedEx',
      'dhl': 'DHL',
      'usps': 'USPS',
      'canada_post': 'Canada Post',
      'purolator': 'Purolator'
    };
    
    return names[carrier] || carrier;
  }

  /**
   * Format tracking status for display
   */
  static formatStatus(status: string): { text: string; color: string; icon: string } {
    const statusMap: Record<string, { text: string; color: string; icon: string }> = {
      'pending': { text: 'Pending', color: 'text-gray-500', icon: 'clock' },
      'in_transit': { text: 'In Transit', color: 'text-blue-500', icon: 'truck' },
      'out_for_delivery': { text: 'Out for Delivery', color: 'text-orange-500', icon: 'map-pin' },
      'delivered': { text: 'Delivered', color: 'text-green-500', icon: 'check-circle' },
      'delayed': { text: 'Delayed', color: 'text-red-500', icon: 'alert-circle' },
      'exception': { text: 'Exception', color: 'text-red-600', icon: 'x-circle' },
      'returned': { text: 'Returned', color: 'text-yellow-600', icon: 'rotate-ccw' },
      'cancelled': { text: 'Cancelled', color: 'text-gray-400', icon: 'x' }
    };
    
    return statusMap[status] || { text: status, color: 'text-gray-500', icon: 'help-circle' };
  }

  /**
   * Format shipment type for display
   */
  static formatShipmentType(type: string): string {
    const types: Record<string, string> = {
      'standard': 'Standard Shipping',
      'express': 'Express Shipping',
      'overnight': 'Overnight Shipping',
      'international': 'International Shipping',
      'freight': 'Freight Shipping'
    };
    
    return types[type] || type;
  }

  /**
   * Get estimated delivery time remaining
   */
  static getTimeRemaining(estimatedDelivery: string): string {
    if (!estimatedDelivery) return 'Unknown';
    
    const now = new Date();
    const delivery = new Date(estimatedDelivery);
    const diffMs = delivery.getTime() - now.getTime();
    
    if (diffMs <= 0) return 'Delivered';
    
    const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));
    const diffHours = Math.floor((diffMs % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
    
    if (diffDays > 0) {
      return `${diffDays} day${diffDays > 1 ? 's' : ''}`;
    } else if (diffHours > 0) {
      return `${diffHours} hour${diffHours > 1 ? 's' : ''}`;
    } else {
      return 'Less than an hour';
    }
  }

  /**
   * Check if shipment is delayed
   */
  static isDelayed(shipment: ShipmentTracking): boolean {
    if (!shipment.estimated_delivery) return false;
    
    const now = new Date();
    const delivery = new Date(shipment.estimated_delivery);
    
    return now > delivery && shipment.status !== 'delivered';
  }

  /**
   * Get latest tracking event
   */
  static getLatestEvent(shipment: ShipmentTracking): TrackingEvent | null {
    if (!shipment.tracking_events || shipment.tracking_events.length === 0) {
      return null;
    }
    
    return shipment.tracking_events.reduce((latest, event) => {
      return new Date(event.event_timestamp) > new Date(latest.event_timestamp) ? event : latest;
    });
  }

  /**
   * Get tracking timeline sorted by date
   */
  static getTrackingTimeline(shipment: ShipmentTracking): TrackingEvent[] {
    if (!shipment.tracking_events) return [];
    
    return [...shipment.tracking_events].sort((a, b) => 
      new Date(b.event_timestamp).getTime() - new Date(a.event_timestamp).getTime()
    );
  }
}

export default ShippingTrackingAPI;

File: backend/models/shipping_tracking.py
"""
Shipping Tracking Models
Integrates with multiple shipping companies (UPS, Canada Express, Royal Mail, etc.)
"""

from sqlalchemy import Column, String, Boolean, ForeignKey, DateTime, Text, Integer, Float, Index
from sqlalchemy.dialects.postgresql import UUID, ENUM as PG_ENUM, JSONB
from sqlalchemy.orm import relationship
from core.db import BaseModel, GUID
from enum import Enum
from datetime import datetime, timezone
from typing import Dict, Any, Optional

class ShippingCarrier(str, Enum):
    """Supported shipping carriers"""
    UPS = "ups"
    CANADA_EXPRESS = "canada_express"
    ROYAL_MAIL = "royal_mail"
    FEDEX = "fedex"
    DHL = "dhl"
    USPS = "usps"
    CANADA_POST = "canada_post"
    PUROLATOR = "purolator"
    TNT = "tnt"
    ARAMEX = "aramex"

    # Added carriers
    LASERSHIP = "lasership"
    ONTRAC = "ontrac"
    HERMES = "hermes"
    EVRI = "evri"              # UK (formerly Hermes)
    DPD = "dpd"
    DPD_LOCAL = "dpd_local"
    GLS = "gls"
    POSTNL = "postnl"
    BPOST = "bpost"
    SWISS_POST = "swiss_post"
    AUSTRALIA_POST = "australia_post"
    NZ_POST = "nz_post"
    JAPAN_POST = "japan_post"
    KOREA_POST = "korea_post"
    CHINA_POST = "china_post"
    SF_EXPRESS = "sf_express"
    YANWEN = "yanwen"
    CAINIAO = "cainiao"
    LAPOSTE = "laposte"
    COLISSIMO = "colissimo"
    CORREOS = "correos"
    POSTE_ITALIANE = "poste_italiane"
    POSTNORD = "postnord"
    BRING = "bring"
    BLUE_DART = "blue_dart"
    DELHIVERY = "delhivery"
    DTDC = "dtdc"
    XPRESSBEES = "xpressbees"

    OTHER = "other"


class TrackingStatus(str, Enum):
    """Tracking status levels"""
    PENDING = "pending"
    IN_TRANSIT = "in_transit"
    OUT_FOR_DELIVERY = "out_for_delivery"
    DELIVERED = "delivered"
    DELAYED = "delayed"
    EXCEPTION = "exception"
    RETURNED = "returned"
    CANCELLED = "cancelled"

class ShipmentType(str, Enum):
    """Types of shipments"""
    STANDARD = "standard"
    EXPRESS = "express"
    OVERNIGHT = "overnight"
    INTERNATIONAL = "international"
    FREIGHT = "freight"

class ShippingProvider(BaseModel):
    """Shipping provider configuration"""
    __tablename__ = "shipping_providers"
    __table_args__ = (
        Index('idx_shipping_providers_carrier', 'carrier'),
        Index('idx_shipping_providers_active', 'is_active'),
        {'extend_existing': True}
    )

    name = Column(String(100), nullable=False)
    carrier = Column(PG_ENUM(ShippingCarrier, name="shipping_carrier"), nullable=False)
    api_key = Column(String(255), nullable=True)  # Encrypted in production
    api_secret = Column(String(255), nullable=True)  # Encrypted in production
    api_url = Column(String(255), nullable=False)
    tracking_url_template = Column(String(500), nullable=False)  # Template for tracking URLs
    webhook_url = Column(String(255), nullable=True)
    is_active = Column(Boolean, default=True)
    configuration = Column(JSONB, nullable=True)  # Provider-specific config
    rate_limits = Column(JSONB, nullable=True)  # API rate limiting config

    # Relationships
    shipments = relationship("ShipmentTracking", back_populates="provider")

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "name": self.name,
            "carrier": self.carrier.value,
            "api_url": self.api_url,
            "tracking_url_template": self.tracking_url_template,
            "is_active": self.is_active,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

class ShipmentTracking(BaseModel):
    """Main shipment tracking model"""
    __tablename__ = "shipment_tracking"
    __table_args__ = (
        Index('idx_shipment_tracking_order_id', 'order_id'),
        Index('idx_shipment_tracking_carrier', 'carrier'),
        Index('idx_shipment_tracking_tracking_number', 'tracking_number'),
        Index('idx_shipment_tracking_status', 'status'),
        Index('idx_shipment_tracking_created_at', 'created_at'),
        Index('idx_shipment_tracking_estimated_delivery', 'estimated_delivery'),
        {'extend_existing': True}
    )

    # Core shipment information
    order_id = Column(GUID(), ForeignKey("orders.id"), nullable=False)
    order_item_id = Column(GUID(), ForeignKey("order_items.id"), nullable=True)  # For multi-item shipments
    provider_id = Column(GUID(), ForeignKey("shipping_providers.id"), nullable=False)
    
    # Tracking details
    tracking_number = Column(String(100), nullable=False, unique=True)
    carrier = Column(PG_ENUM(ShippingCarrier, name="shipment_carrier"), nullable=False)
    status = Column(PG_ENUM(TrackingStatus, name="tracking_status"), default=TrackingStatus.PENDING)
    shipment_type = Column(PG_ENUM(ShipmentType, name="shipment_type"), default=ShipmentType.STANDARD)
    
    # Timeline information
    shipped_at = Column(DateTime(timezone=True), nullable=True)
    estimated_delivery = Column(DateTime(timezone=True), nullable=True)
    actual_delivery = Column(DateTime(timezone=True), nullable=True)
    
    # Location information
    origin_address = Column(JSONB, nullable=True)  # Pickup address
    destination_address = Column(JSONB, nullable=True)  # Delivery address
    current_location = Column(JSONB, nullable=True)  # Current location
    delivery_instructions = Column(Text, nullable=True)
    
    # Package information
    package_weight = Column(Float, nullable=True)  # Weight in kg
    package_dimensions = Column(JSONB, nullable=True)  # {length, width, height} in cm
    package_value = Column(Float, nullable=True)  # Declared value
    insurance_amount = Column(Float, nullable=True)
    
    # Service details
    service_level = Column(String(50), nullable=True)  # Express, Standard, etc.
    delivery_signature_required = Column(Boolean, default=False)
    delivery_confirmation = Column(String(100), nullable=True)
    
    # External tracking data
    external_tracking_data = Column(JSONB, nullable=True)  # Raw data from carrier API
    last_api_sync = Column(DateTime(timezone=True), nullable=True)
    sync_status = Column(String(50), default="pending")  # pending, success, error
    
    # Customer notifications
    customer_notified = Column(Boolean, default=False)
    notification_preferences = Column(JSONB, nullable=True)
    
    # Metadata
    notes = Column(Text, nullable=True)
    internal_notes = Column(Text, nullable=True)

    # Relationships
    order = relationship("Order", back_populates="shipments")
    order_item = relationship("OrderItem", back_populates="shipment")
    provider = relationship("ShippingProvider", back_populates="shipments")
    tracking_events = relationship("ShipmentTrackingEvent", back_populates="shipment", cascade="all, delete-orphan")

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "order_id": str(self.order_id),
            "tracking_number": self.tracking_number,
            "carrier": self.carrier.value,
            "status": self.status.value,
            "shipment_type": self.shipment_type.value,
            "shipped_at": self.shipped_at.isoformat() if self.shipped_at else None,
            "estimated_delivery": self.estimated_delivery.isoformat() if self.estimated_delivery else None,
            "actual_delivery": self.actual_delivery.isoformat() if self.actual_delivery else None,
            "current_location": self.current_location,
            "delivery_instructions": self.delivery_instructions,
            "package_weight": self.package_weight,
            "package_dimensions": self.package_dimensions,
            "service_level": self.service_level,
            "delivery_signature_required": self.delivery_signature_required,
            "external_tracking_url": self.get_tracking_url(),
            "tracking_events": [event.to_dict() for event in self.tracking_events],
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

    def get_tracking_url(self) -> Optional[str]:
        """Generate carrier-specific tracking URL"""
        if not self.provider or not self.tracking_number:
            return None
            
        template = self.provider.tracking_url_template
        if not template:
            return None
            
        return template.replace("{tracking_number}", self.tracking_number)

class ShipmentTrackingEvent(BaseModel):
    """Individual tracking events for a shipment"""
    __tablename__ = "shipment_tracking_events"
    __table_args__ = (
        Index('idx_tracking_events_shipment_id', 'shipment_id'),
        Index('idx_tracking_events_timestamp', 'event_timestamp'),
        Index('idx_tracking_events_event_type', 'event_type'),
        {'extend_existing': True}
    )

    shipment_id = Column(GUID(), ForeignKey("shipment_tracking.id"), nullable=False)
    event_timestamp = Column(DateTime(timezone=True), nullable=False)
    event_type = Column(String(50), nullable=False)  # picked_up, in_transit, out_for_delivery, delivered, etc.
    event_description = Column(Text, nullable=False)
    event_location = Column(JSONB, nullable=True)  # {city, state, country, coordinates}
    
    # Carrier-specific data
    carrier_event_code = Column(String(50), nullable=True)
    carrier_event_data = Column(JSONB, nullable=True)
    
    # Additional details
    estimated_delivery = Column(DateTime(timezone=True), nullable=True)
    delay_reason = Column(String(255), nullable=True)
    exception_details = Column(JSONB, nullable=True)
    
    # Contact information
    contact_name = Column(String(100), nullable=True)
    contact_phone = Column(String(20), nullable=True)
    
    # Metadata
    source = Column(String(50), default="api")  # api, webhook, manual
    raw_data = Column(JSONB, nullable=True)

    # Relationships
    shipment = relationship("ShipmentTracking", back_populates="tracking_events")

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "shipment_id": str(self.shipment_id),
            "event_timestamp": self.event_timestamp.isoformat(),
            "event_type": self.event_type,
            "event_description": self.event_description,
            "event_location": self.event_location,
            "carrier_event_code": self.carrier_event_code,
            "estimated_delivery": self.estimated_delivery.isoformat() if self.estimated_delivery else None,
            "delay_reason": self.delay_reason,
            "contact_name": self.contact_name,
            "contact_phone": self.contact_phone,
            "source": self.source,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }

class ShippingWebhook(BaseModel):
    """Webhook configurations for shipping updates"""
    __tablename__ = "shipping_webhooks"
    __table_args__ = (
        Index('idx_shipping_webhooks_provider', 'provider_id'),
        Index('idx_shipping_webhooks_active', 'is_active'),
        {'extend_existing': True}
    )

    provider_id = Column(GUID(), ForeignKey("shipping_providers.id"), nullable=False)
    webhook_url = Column(String(500), nullable=False)
    webhook_secret = Column(String(255), nullable=True)
    event_types = Column(JSONB, nullable=True)  # Which events to trigger on
    is_active = Column(Boolean, default=True)
    retry_count = Column(Integer, default=0)
    last_triggered = Column(DateTime(timezone=True), nullable=True)
    success_count = Column(Integer, default=0)
    error_count = Column(Integer, default=0)

    # Relationships
    provider = relationship("ShippingProvider")

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": str(self.id),
            "provider_id": str(self.provider_id),
            "webhook_url": self.webhook_url,
            "event_types": self.event_types,
            "is_active": self.is_active,
            "retry_count": self.retry_count,
            "success_count": self.success_count,
            "error_count": self.error_count,
            "last_triggered": self.last_triggered.isoformat() if self.last_triggered else None,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }
